### PR TITLE
@W-16890766:  graphql scaffolding and test coverage. 

### PR DESCRIPTION
**Pull request by ben-zhang-at-salesforce**
[W-16890766](https://gus.lightning.force.com/a07EE000022R15zYAC)

Scaffolding graphql diagnostic producing mechanism.
- pull gql'...` out calling gqlPluckFromCodeStringSync.
- parse the graphql
- call into rules to generate diagnostic,  we did a dummy rule to show case the setup. 

- test coverage

beside above we set the source for diagnostics as 'Salesforce Mobile Extensions' for developer to identify the source of the diagnostics.

### CODE REVIEW

File: lsp/server/src/validateGraphql.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * @param graphqlDiagnosticProducers  the collection of graphql rules. 
+ */
+export async function validateOneGraphQuery(textDocument: TextDocument, graphql: string): Promise<Diagnostic[]> {
+    const results: Diagnostic[] = []; 
+  
+    // graphql string fails to parse will not produce diagnostic
+    try {
+        const graphqlAstNode = parse(graphql);
+
+        for (const producer of diagnosticProducers) {
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/validateGraphql.ts:77]: We can run producer.validateDocument parallel
```suggestion
       const allResults = await Promise.all(
            diagnosticProducers.map((producer) =>
                producer.validateDocument(textDocument, graphqlAstNode)
            )
        );
        return allResults.flat();
```

File: lsp/server/src/validateGraphql.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    );
+
+    // Validate each query
+    for (const query of graphQueries) {
+        if (results.length >= maxCount) {
+            break;
+        }
+        const lineOffset = query.locationOffset.line - 1;
+        const columnOffset = query.locationOffset.column + 1;
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/validateGraphql.ts:49]: Why is column offset +1?

File: lsp/server/src/validateGraphql.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import { Diagnostic } from 'vscode-languageserver/node';
+import { DiagnosticProducer } from './diagnostic/DiagnosticProducer';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { MisspelledUiapi } from './diagnostic/gql/misspelled-uiapi';
+
+const diagnosticProducers: DiagnosticProducer<ASTNode>[] = [];
+diagnosticProducers.push(new MisspelledUiapi());
+
+/**
+ * validate the graphql queries in the document.
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/validateGraphql.ts:20]: Nit: capitalize

File: lsp/server/src/validateGraphql.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if (maxCount <= 0 || diagnosticProducers.length === 0) {
+        return results;
+    }
+
+    // Find the gql``s in the file content
+    const graphQueries = gqlPluckFromCodeStringSync(
+        textDocument.uri,
+        textDocument.getText(),
+        {
+            skipIndent: true
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/validateGraphql.ts:39]: Probably add safeguard for plucker
```suggestion
         
            {
            skipIndent: true,
            globalGqlIdentifierName: ['gql', 'graphql']
          }
```

File: lsp/server/src/validateGraphql.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    try {
+        const graphqlAstNode = parse(graphql);
+
+        for (const producer of diagnosticProducers) {
+            (await producer.validateDocument(textDocument, graphqlAstNode)).forEach((it) => {
+                results.push(it);
+            });
+        }
+    } catch (e) {
+        // Do nothing here?
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/validateGraphql.ts:83]: Nit: remove it and explain reason of error silence

File: lsp/server/tsconfig.json [hunk 6]
```diff
@@ -11,6 +11,6 @@
         "composite": true
     },
 
-    "include": ["src"],
+    "include": ["src", "src/diagnostic/gql/misspelled-uiapi.ts"],
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/tsconfig.json:14]: Do we need to include `src/diagnostic/gql/misspelled-uiapi.ts`

File: package.json [hunk 7]
```diff
@@ -90,8 +90,7 @@
         "prettier:write": "prettier --write \"src/**/*.{ts, js}\" \"resources/instructions/**/*.html\"",
         "prettier:verify": "prettier --list-different \"src/**/*.{ts, js}\" \"resources/instructions/**/*.html\"",
         "bundle:extension": "esbuild ./src/extension.ts --bundle --outdir=out --format=cjs --target=es2020 --platform=node --external:vscode --external:@salesforce/core --external:@oclif/core --external:@salesforce/lwc-dev-mobile-core --minify --sourcemap",
-        "vscode:prepublish": "npm run clean && npm run bundle:extension",
-        "postinstall": "cd lsp/client && npm install && cd ../server && npm install && cd ../.."
```
Comment (haifeng-li-at-salesforce) [inline package.json]: If we remove 'postinstall', the subprojects will not install their own modules

File: lsp/server/src/validateGraphql.ts (hunk 2 above)
Comment (ben-zhang-at-salesforce) [reply lsp/server/src/validateGraphql.ts:49]: graphql-eslint has same code.  seems that 

- query.locationOffset's line and column is 1 based. 
- query.locationOffset.column value starts at the `. so need to plus 1 to get the correct body start position.

File: package.json (hunk 7 above)
Comment (ben-zhang-at-salesforce) [reply package.json]: ya, my mistake.

File: lsp/server/tsconfig.json (hunk 6 above)
Comment (ben-zhang-at-salesforce) [reply lsp/server/tsconfig.json:14]: probably not, let me try removing.

**Review by haifeng-li-at-salesforce — APPROVED** [review]
Nice job!

**Pull request merged**